Fixes #1599 - Improve the Function apps browse list

agruning opened the pull request:
![image](https://user-images.githubusercontent.com/8560637/28798605-da440cb8-75f9-11e7-9248-080729f57e5e.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/azure/azure-functions-ux/1611)
<!-- Reviewable:end -->

dnfclas commented:
@agruning,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

ahmelsayed commented:
the use of `.forEach(.push)` pattern is a bit weird. Js has few functional constructs like `.map`, `.filter`, `.reduce`, etc.

for example this pattern (_pseudocode._)

```ts
const newApps: App[] = [];

this.apps.forEach(a => {
    if(a.blah) {
        newItems.push(a);
    }
});

this.apps = newItems;
```

can be converted to

```ts
this.apps = this.apps.filter(a => a.blah);
```

and this pattern

```ts
const anotherTable = [];

this.apps.forEach(a => {
    if(a.blah) {
        anotherTable.push({
            app: a,
            something: "something else"
        });
    }
});

```

can be converted to 

```ts
const anotherTable = this.apps
    .filter(a => a.blah)
    .map(a => ({app: a, something: "something else"}));
```

almost any use for `forEach`, `for...of`, `.push`, etc could be replaced with one of the generic functional constructs above. It makes it much easier to read.

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -1,5 +1,12 @@
+import { ResourceGroup } from './../shared/models/resource-group';
+import { element } from 'protractor';
```
not used?

agruning replied:
fixed

agruning replied:
removed

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -51,7 +92,82 @@ export class AppsListComponent implements OnInit, OnDestroy {
+    item.appNode.sideNav.searchExact(item.title);
+  }
+
+  contains(array: any[], element: any) {
```
this method is equivalent to [`Array.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find), we should use that instead.

agruning replied:
updated

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -51,7 +92,82 @@ export class AppsListComponent implements OnInit, OnDestroy {
+        this._updateLocDisplayText(this.translateService.instant(PortalResources.locationCount).format(locations.length));
+    } else {
+        this._updateLocDisplayText(`${locations[0]}`);
+    }
```
can you add a comment explaining what these conditions are for? something like
```ts
// if all locations are selected

// else if one or more locations are selected

// etc
```

agruning replied:
done

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -51,7 +92,82 @@ export class AppsListComponent implements OnInit, OnDestroy {
+
+    if (this.selectedLocations.length === this.locationOptions.length) {
+        this._updateLocDisplayText(this.translateService.instant(PortalResources.allLocations));
+    } else if (this.selectedLocations.length > 1) {
```
what if `this.selectedLocations.length === 1`?

agruning replied:
That is the else case. I modeled this if/else if/else statement after the dropdown case used in side-nav.component

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -51,7 +92,82 @@ export class AppsListComponent implements OnInit, OnDestroy {
+    this.locationsDisplayText = '';
+    setTimeout(() => {
+        this.locationsDisplayText = displayText;
+    }, 10);
```
why the setTimeout?

agruning replied:
The page has to be re-rendered for the updated display text to display. I have changed the time to 0.

ahmelsayed commented on AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts:
```diff
@@ -49,23 +49,63 @@ export class TblThComponent implements OnInit {
+        }
+      });
+    }
+    else {
```
nit: formatting, put else on the previous line.

agruning replied:
done

ahmelsayed commented on AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts:
```diff
@@ -49,23 +49,63 @@ export class TblThComponent implements OnInit {
+        if (table.sortAscending) {
+          return aCol.localeCompare(bCol);
+        }
+        else {
```
nit: formatting, put else on the previous line

agruning replied:
done

ahmelsayed commented on AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts:
```diff
@@ -49,23 +49,63 @@ export class TblThComponent implements OnInit {
+
+        tempItems.sort((a: TableItem, b: TableItem) => {
+          if (table.sortAscending) {
+            return a[this.name].localeCompare(b[this.name]);
```
are `a[this.name]` and `b[this.name]` both guaranteed to exist?

agruning replied:
Yes, this code was already in-place for sorting without grouping. All elements sorted by this method with have this component.

ahmelsayed commented on AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts:
```diff
@@ -315,4 +322,67 @@ export class TblComponent implements OnInit, OnChanges {
     return this._origItems;
   }
+
+  contains(array: any[], element: any) {
```
use `.find` instead as mentioned before.

agruning replied:
done

ehamai commented:
Reviewed 1 of 9 files at r1.
Review status: 1 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqP_28_fHcnxTC33pSC:-KqP_28ajoHthkaUSGvz:b-jc8yb9) (no related file):*
Haven't looked at the code yet, but the UI looks very nice!  Can you also please make sure that keyboard navigation still works?

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 20 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPbRN-6DFL7e36SCkI:-KqPbRN-6DFL7e36SCkJ:b-oghrjq) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L20)):*
> ```HTML
> </div> 
> 
> <tbl [items]="tableItems" #table="tbl" id="apps-list" [name]='Resources.functionApps | translate'>
> ```

I think you can just do [name]="'functionApps' | translate"

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 34 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPbm0V13eOHnvkB0EC:-KqPbm0V13eOHnvkB0ED:bj23nyn) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L34)):*
> ```HTML
>     <td *ngIf="item.type === 'row' && table.groupedBy === 'none'">{{item.location}}</td>
>     
>     <td class='tabbed' *ngIf="item.type === 'row' && table.groupedBy !== 'none'"><a (click)="clickRow(item)">{{item.title}}</a></td>
> ```

I just learned about using ng-container from Aniello.  Basically you should be able to wrap your <td> elements with a single <ng-container *ngIf="item.type === 'row' && table.groupedBy === 'none'"> element.  When Angular renders the DOM, it will remove the ng-container element but apply the same view logic to the internal elements like you're doing.  This will help to keep your HTML clean.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 39 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPd1Lo_bbocCc_PcgZ:-KqPd1Lo_bbocCc_Pcg_:b-ngkw8q) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L39)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     <td *ngIf="item.type === 'group'">{{item.title}}</td>
>     <td *ngIf="item.type === 'group'"></td>
>     <td *ngIf="item.type === 'group'"></td>
>     <td *ngIf="item.type === 'group'"></td>
</details>

Instead of defining 3 td elements, just use colspan on the first

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ahmelsayed commented on AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts:
```diff
@@ -315,4 +322,67 @@ export class TblComponent implements OnInit, OnChanges {
+      }
+    }
+
+    this.items = newItems;
```
can be simplified to (_I think_)

```ts
groupItems(name: string) {
     this.groupedBy = name;
 
    this.items = this.items
        .filter(i => i.type !== 'group')
        .sort((a: TableItem, b: TableItem) => b.title.localeCompare(a.title));
```

though I have no idea what the filtering on `newItems[index].type === 'group'` is trying to do.

ahmelsayed commented on AzureFunctions.AngularClient/src/app/functions-list/functions-list.component.html:
```diff
@@ -20,7 +20,7 @@
         <tr>
             <th>
-                <tbl-th name="title">{{ '_name' | translate }}</tbl-th>
+                <tbl-th name="functionInfo.name">{{ '_name' | translate }}</tbl-th>
```
can `functionInfo` be null?

ahmelsayed commented:
Review status: 1 of 9 files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqP_28_fHcnxTC33pSC:-KqPeOJaY3AFt5KRTRtb:b-hmkfpv) (no related file):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

Haven't looked at the code yet, but the UI looks very nice!  Can you also please make sure that keyboard navigation still works?
</blockquote></details>

yep, the UI looks great. I just have a few comments about the pattern handling Array operations. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ahmelsayed commented on AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts:
```diff
@@ -17,14 +24,32 @@ import { TreeViewInfo } from './../tree-view/models/tree-view-info';
+  @ViewChild('nameHeader') nameHeader: TblThComponent;
+  @ViewChild('subHeader') subHeader: TblThComponent;
+  @ViewChild('resHeader') resHeader: TblThComponent;
+  @ViewChild('locHeader') locHeader: TblThComponent;
```
what are these used for?

agruning replied:
No longer needed. I have removed

ehamai commented:
Review status: 1 of 9 files reviewed at latest revision, 17 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 99 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPZsoF719pSnh_7gMq-r1-99:-KqPfBcqufVXRMkeHxKz:b-l5m9rn) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L99)):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

this method is equivalent to [`Array.find`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find), we should use that instead.
</blockquote></details>

Hmmm.  Dunno how to give thumbs up in this tool.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 141 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPfcz91Cdwfywbi2Wr:-KqPfcz91Cdwfywbi2Ws:b-qv4nq4) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L141)):*
> setTimeout(() => {
>         this.appTable.groupItems(this.currGroup);
>     }, 0);
Add a comment on why this is needed

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 147 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPZsoF719pSnh_7gMq-r1-147:-KqPflabua3mvnxHRP9U:bdqz7w6) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L147)):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

what if `this.selectedLocations.length === 1`?
</blockquote></details>

Looks like it would just return the first location in that case

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ahmelsayed commented:
Review status: 1 of 9 files reviewed at latest revision, 17 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 147 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPZsoF719pSnh_7gMq-r1-147:-KqPiOVECQcXGsNTybMD:b-vf5ng7) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L147)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

Looks like it would just return the first location in that case
</blockquote></details>

but shouldn't it return `this.selectedLocations[0]` then? I'm a bit confused about `locations` vs `this.selectedLocations` vs `this.locationOptions` 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ehamai commented:
Review status: 1 of 9 files reviewed at latest revision, 22 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 168 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPg4tgjzxb1AesLayw:-KqPg4tgjzxb1AesLayx:bpcwle9) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L168)):*
> setTimeout(() => {
>         this.appTable.groupItems(group);
>     }, 0);

Add comment on why this is necessary

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 9 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPgBliW-tiawG_WkCP:-KqPgBliW-tiawG_WkCQ:bshe7wk) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L9)):*
> export interface TableItem {
>   title: string;
>   subscription: string;
>   resourceGroup: string;
>   location: string;
>   type: 'row' | 'group';
>   appNode?: AppNode;
> }

We shouldn't have anything that's not generic in the TableItem interface.  Maybe a better way to do this would be something like this:

export interface TableItem{
   type: 'row' | 'group';
}

Then for appsList, we could create a more specific interface:

interface AppItem extends TableItem{
   title: string;
   resourceGroup: string;
   ...
}

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 340 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPhBX_CocmnyNFaj6Q:-KqPhBX_CocmnyNFaj6R:b-fbtk41) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L340)):*
> return b.title.localeCompare(a.title);

I think ideally we shouldn't assume that there will be a property called "title".  That column name should be passed in by the caller.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 354 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjskrX7vABnRLX0SO:-KqPjskrX7vABnRLX0SP:b-pbtm0g) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L354)):*
> if (!this.contains(uniqueGroups, item[name])) {

Instead of always checking for existence, why don't you just keep assigning to an object property with the group name.  This would save you many loops:

uniqueGroups[name] = item[name];

You'd have to get rid of the newItems.push below, but you can move that into when you do your final iteration for finalItems

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 369 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjZ-IdMd0Vhu4hCBH:-KqPjZ-IdMd0Vhu4hCBI:b-2eyg9k) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L369)):*
> newItems.reverse();
What is the reverse for?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ehamai commented:
Review status: 1 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 75 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPkiKBv99qjeFxSKgO:-KqPkiKBv99qjeFxSKgP:b-vmqedp) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L75)):*
> appItems
nit: please rename to something generic

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 94 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPls-Exu_hnRzu7LTf:-KqPls-Exu_hnRzu7LTg:bci976w) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L94)):*
> tempItems.sort((a: TableItem, b: TableItem) => {

I think ideally we should be doing sorted inserts which would be faster than calling sort on each group.  But for now let's just leave it.  If we get feedback on performance then we can think about how to optimize.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ahmelsayed commented:
@agruning 
Actually I take back my comment about a possible compiler bug. I think it actually makes sense.

when you do

```ts
interface Item {
   kind: 'row';
   id: number;
}

const arr: Item[] = [1, 2, 3].map(id => ({id: id, kind: 'row'}));
```

it seems logical that map result should [duck ](https://en.wikipedia.org/wiki/Duck_typing) match `Item` interface, but the compiler just isn't that smart to understand that `kind` from the map can only be of type `'row'` but rather a generic `string`. I don't think it's a compiler bug. this does seem like the right solution to me.

```ts
const arr: Item[] = [1, 2, 3].map(id => <Item> ({id: id, kind: 'row'}));

// Or

const arr: Item[] = [1, 2, 3].map(id => ({id: id, kind: 'row'} as Item));
```

agruning commented:
Review status: 1 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqP_28_fHcnxTC33pSC:-KqPh0nYHiO24nJQgIKA:b-fz6qm7) (no related file):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

yep, the UI looks great. I just have a few comments about the pattern handling Array operations. 
</blockquote></details>

Keyboard navigation works! And thanks for the array pattern improvements. I'll get those updated.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 20 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPbRN-6DFL7e36SCkI:-KqPtXlwGpXJBEVIVZQI:bb6un1j) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L20)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

I think you can just do [name]="'functionApps' | translate"
</blockquote></details>

Done

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 34 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPbm0V13eOHnvkB0EC:-KqPu3KeG08Dy0CNGdws:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L34)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

I just learned about using ng-container from Aniello.  Basically you should be able to wrap your <td> elements with a single <ng-container *ngIf="item.type === 'row' && table.groupedBy === 'none'"> element.  When Angular renders the DOM, it will remove the ng-container element but apply the same view logic to the internal elements like you're doing.  This will help to keep your HTML clean.
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html, line 39 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPd1Lo_bbocCc_PcgZ:-KqPu3l2Rr1FhbtN5inN:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.html#L39)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

>     <td *ngIf="item.type === 'group'">{{item.title}}</td>
>     <td *ngIf="item.type === 'group'"></td>
>     <td *ngIf="item.type === 'group'"></td>
>     <td *ngIf="item.type === 'group'"></td>

Instead of defining 3 td elements, just use colspan on the first
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 141 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPfcz91Cdwfywbi2Wr:-KqPu6WxJwo_PDDuWx4A:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L141)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> setTimeout(() => {
>         this.appTable.groupItems(this.currGroup);
>     }, 0);
Add a comment on why this is needed
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 168 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPg4tgjzxb1AesLayw:-KqPu7l3ggA8Uh6TTmC8:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L168)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> setTimeout(() => {
>         this.appTable.groupItems(group);
>     }, 0);

Add comment on why this is necessary
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 340 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPhBX_CocmnyNFaj6Q:-KqQ4lSYHDOiGwHQYGBG:b-w1ilaz) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L340)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> return b.title.localeCompare(a.title);

I think ideally we shouldn't assume that there will be a property called "title".  That column name should be passed in by the caller.
</blockquote></details>

This is when we initialize groups into the table. When this happens, by default each group is sorted by its title. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 75 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPkiKBv99qjeFxSKgO:-KqQ7FXgEh3uWB2tlakY:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L75)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> appItems
nit: please rename to something generic
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 369 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjZ-IdMd0Vhu4hCBH:-KqQ7mwPNfGLIpLDLuhx:b-by033w) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L369)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> newItems.reverse();
What is the reverse for?
</blockquote></details>

This reverse makes it so all group type items are listed before all row type items.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 9 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPgBliW-tiawG_WkCP:-KqTcTOzz0Wr-JMHCC5N:bb6un1j) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L9)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> export interface TableItem {
>   title: string;
>   subscription: string;
>   resourceGroup: string;
>   location: string;
>   type: 'row' | 'group';
>   appNode?: AppNode;
> }

We shouldn't have anything that's not generic in the TableItem interface.  Maybe a better way to do this would be something like this:

export interface TableItem{
   type: 'row' | 'group';
}

Then for appsList, we could create a more specific interface:

interface AppItem extends TableItem{
   title: string;
   resourceGroup: string;
   ...
}
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 349 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPZsoF719pSnh_7gMs-r1-349:-KqThd0o95_G25U0_h_F:b-dlnt20) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L349)):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

can be simplified to (_I think_)

```ts
groupItems(name: string) {
     this.groupedBy = name;
 
    this.items = this.items
        .filter(i => i.type !== 'group')
        .sort((a: TableItem, b: TableItem) => b.title.localeCompare(a.title));
```

though I have no idea what the filtering on `newItems[index].type === 'group'` is trying to do.
</blockquote></details>

Updated. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 94 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPls-Exu_hnRzu7LTf:-KqQ9B403jOPkg9MEf5l:b-hfghm4) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L94)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> tempItems.sort((a: TableItem, b: TableItem) => {

I think ideally we should be doing sorted inserts which would be faster than calling sort on each group.  But for now let's just leave it.  If we get feedback on performance then we can think about how to optimize.
</blockquote></details>

Sounds good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ahmelsayed commented:
Review status: 0 of 9 files reviewed at latest revision, 24 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 369 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjZ-IdMd0Vhu4hCBH:-KqUd6omo24ALTGpXmMz:b8athlc) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L369)):*
<details><summary><i>Previously, agruning (Allison Gruninger) wrote…</i></summary><blockquote>

This reverse makes it so all group type items are listed before all row type items.
</blockquote></details>

I'd add your reply as a comment explaining why that's needed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/functions-list/functions-list.component.html, line 23 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPZsoF719pSnh_7gMt-r1-23:-KqUcNbTzbpIimpBBoPa:b16y832) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/functions-list/functions-list.component.html#L23)):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

can `functionInfo` be null?
</blockquote></details>

No, all function-node elements must have functionInfo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 369 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjZ-IdMd0Vhu4hCBH:-KqUe3p-Cb8XCuWD-OkO:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L369)):*
<details><summary><i>Previously, ahmelsayed (Ahmed ElSayed) wrote…</i></summary><blockquote>

I'd add your reply as a comment explaining why that's needed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 354 at r1](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqPjskrX7vABnRLX0SO:-KqUe99oxeqh3b8GYRfz:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/e78f0dc72570818193a30fdbfdfe08dca7796792/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L354)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> if (!this.contains(uniqueGroups, item[name])) {

Instead of always checking for existence, why don't you just keep assigning to an object property with the group name.  This would save you many loops:

uniqueGroups[name] = item[name];

You'd have to get rid of the newItems.push below, but you can move that into when you do your final iteration for finalItems
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ehamai commented:
Reviewed 5 of 9 files at r1, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 66 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYb3ph0FPAy26JWI-j:-KqYb3ph0FPAy26JWI-k:bl8tuh7) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L66)):*
>  this.tableItems = this.apps.map(app => (<AppTableItem>{title: app.title,
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>
>                                               subscription: app.subscription,
>                                               type: 'row',
>                                               resourceGroup: app.resourceGroup,
>                                               location: app.location,
>                                               appNode: app}));
</details>

nit: formatting looks a little weird here

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 109 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYfGZqrewosvDtmpdH:-KqYfGZqrewosvDtmpdI:b-cxj2cy) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L109)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>
>                               .map(app => (<AppTableItem> {title: app.title,
>                                                         subscription: app.subscription,
>                                                         type: 'row',
>                                                         resourceGroup: app.resourceGroup,
>                                                         location:  this.translateService.instant(app.location),
>                                                         appNode: app})));
</details>

nit: formatting looks a little weird

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 32 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYdhsVx197hEN-yPzQ:-KqYdhsVx197hEN-yPzR:b-q05va5) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L32)):*
> @Input() groupColName: string | null;

Add a comment on what this means, as well as what groupedBy means.  They will likely get confused by someone modifying this component

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 349 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYma-urDpTZEebaE2D:-KqYma-urDpTZEebaE2E:b-dtyzp5) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L349)):*
>      // determine uniqueGroup value

This should be slightly faster to build unique groups:

        let newItems = [];
        let uniqueGroups = {};
        newItems.forEach(item =>{
            uniqueGroups[item[name]] = item[name];
        })

        let groups = [];
        for(let group in uniqueGroups){
            if(uniqueGroups.hasOwnProperty(group)){
                groups.push(group);
            }
        }

        groups.sort();

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 76 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYcmDkswKkH7ucb9Fn:-KqYcmDkswKkH7ucb9Fo:budm8t) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L76)):*
>       const groupItems = table.items.filter(item => item.type === 'group');
>       const rowItems = table.items.filter(item => item.type === 'row');

You can get rid of one iteration of the collection by creating groupItems and rowItems in one loop:

table.items.forEach(i => {
   if(item.type === 'group'){
      groupItems.push(item);
   } else if(item.type === 'row'){
      rowItems.push(item);
   }
});

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 81 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYe5GBCipUksgWWdnT:-KqYe5GBCipUksgWWdnU:byozv5t) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L81)):*
> app
Rename to "row" instead of "app"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

agruning commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 66 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYb3ph0FPAy26JWI-j:-KqYr5ISaf-i8_miYI33:b-bw8e14) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L66)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

>  this.tableItems = this.apps.map(app => (<AppTableItem>{title: app.title,
>                                               subscription: app.subscription,
>                                               type: 'row',
>                                               resourceGroup: app.resourceGroup,
>                                               location: app.location,
>                                               appNode: app}));

nit: formatting looks a little weird here
</blockquote></details>

Updated.

---

*[AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts, line 109 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYfGZqrewosvDtmpdH:-KqYoE5fS2kbEDHnfXy8:b-bw8e14) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/apps-list/apps-list.component.ts#L109)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

>                               .map(app => (<AppTableItem> {title: app.title,
>                                                         subscription: app.subscription,
>                                                         type: 'row',
>                                                         resourceGroup: app.resourceGroup,
>                                                         location:  this.translateService.instant(app.location),
>                                                         appNode: app})));

nit: formatting looks a little weird
</blockquote></details>

Updated.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 32 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYdhsVx197hEN-yPzQ:-KqYoIZjn4nNFmpe6_U5:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L32)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> @Input() groupColName: string | null;

Add a comment on what this means, as well as what groupedBy means.  They will likely get confused by someone modifying this component
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts, line 349 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYma-urDpTZEebaE2D:-KqYqPL5AzvzhJPvxGXB:bic8q00) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl.component.ts#L349)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

>      // determine uniqueGroup value

This should be slightly faster to build unique groups:

        let newItems = [];
        let uniqueGroups = {};
        newItems.forEach(item =>{
            uniqueGroups[item[name]] = item[name];
        })

        let groups = [];
        for(let group in uniqueGroups){
            if(uniqueGroups.hasOwnProperty(group)){
                groups.push(group);
            }
        }

        groups.sort();
</blockquote></details>

Updated. Thank you!

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 76 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYcmDkswKkH7ucb9Fn:-KqYr1KKrjpf8JL01RrX:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L76)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

>       const groupItems = table.items.filter(item => item.type === 'group');
>       const rowItems = table.items.filter(item => item.type === 'row');

You can get rid of one iteration of the collection by creating groupItems and rowItems in one loop:

table.items.forEach(i => {
   if(item.type === 'group'){
      groupItems.push(item);
   } else if(item.type === 'row'){
      rowItems.push(item);
   }
});
</blockquote></details>

Done.

---

*[AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts, line 81 at r4](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611#-KqYe5GBCipUksgWWdnT:-KqYr1vb8nx3EkBp7sZh:b-896fix) ([raw file](https://github.com/azure/azure-functions-ux/blob/7e54b2904c6b8c146d27807ec9df960357804fd4/AzureFunctions.AngularClient/src/app/controls/tbl/tbl-th/tbl-th.component.ts#L81)):*
<details><summary><i>Previously, ehamai (Elliott Hamai) wrote…</i></summary><blockquote>

> app
Rename to "row" instead of "app"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->

ehamai commented:
Reviewed 1 of 3 files at r5.
Review status: 7 of 9 files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-functions-ux/1611)*
<!-- Sent from Reviewable.io -->
